File: app/src/main/java/doit/study/droid/adapters/TopicsAdapter.java
package doit.study.droid.adapters;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import doit.study.droid.R;
import doit.study.droid.data.source.Tag;
import timber.log.Timber;

public class TopicsAdapter extends RecyclerView.Adapter<TopicsAdapter.TopicViewHolder>{
    private final static boolean DEBUG = false;
    private List<Tag> mFilteredTags = new ArrayList<>();
    private List<Tag> mMasterCopyTags = new ArrayList<>();


    public void animateTo(List<Tag> models) {
        if (DEBUG) Timber.d("Anim Filtered %d, current: %d", models.size(), mFilteredTags.size());
        applyAndAnimateRemovals(models);
        if (DEBUG) Timber.d("Rem Filtered %d, current: %d", models.size(), mFilteredTags.size());
        applyAndAnimateAdditions(models);
        if (DEBUG) Timber.d("Add Filtered %d, current: %d", models.size(), mFilteredTags.size());
        applyAndAnimateMovedItems(models);
        if (DEBUG) Timber.d("Move Filtered %d, current: %d", models.size(), mFilteredTags.size());
    }

    private void applyAndAnimateRemovals(List<Tag> newTags) {
        for (int i = mFilteredTags.size() - 1; i >= 0; i--){
            final Tag tag = mFilteredTags.get(i);
            if (!newTags.contains(tag)){
                removeItem(i);
            }
        }
    }

    private void applyAndAnimateAdditions(List<Tag> newModels) {
        for (int i = 0, count = newModels.size(); i < count; i++) {
            final Tag model = newModels.get(i);
            if (!mFilteredTags.contains(model)) {
                addItem(i, model);
            }
        }
    }

    private void applyAndAnimateMovedItems(List<Tag> newModels) {
        for (int toPosition = newModels.size() - 1; toPosition >= 0; toPosition--) {
            final Tag model = newModels.get(toPosition);
            final int fromPosition = mFilteredTags.indexOf(model);
            if (fromPosition >= 0 && fromPosition != toPosition) {
                moveItem(fromPosition, toPosition);
            }
        }
    }

    public Tag removeItem(int position) {
        final Tag model = mFilteredTags.remove(position);
        notifyItemRemoved(position);
        notifyItemRangeChanged(position, getItemCount());
        return model;
    }

    public void addItem(int position, Tag model) {
        mFilteredTags.add(position, model);
        notifyItemInserted(position);
    }

    public void moveItem(int fromPosition, int toPosition) {
        final Tag model = mFilteredTags.remove(fromPosition);
        mFilteredTags.add(toPosition, model);
        notifyItemMoved(fromPosition, toPosition);
    }

    public static class TopicViewHolder extends RecyclerView.ViewHolder{
        TextView topic;
        CheckBox checkbox;
        public TopicViewHolder(View itemView) {
            super(itemView);
            topic = itemView.findViewById(R.id.topic_name);
            checkbox = itemView.findViewById(R.id.checkbox_tag);
        }

    }

    @Override
    public int getItemCount() {
        if (mFilteredTags != null)
            return mFilteredTags.size();
        else
            return 0;
    }


    public List<Tag> getTags(){
        return mMasterCopyTags;
    }


    public void setTags(List<Tag> tags){
        mMasterCopyTags = tags;
        mFilteredTags = new ArrayList<>(tags);
        notifyDataSetChanged();
    }

    @Override
    public TopicViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_topic_item, parent, false);
        return new TopicViewHolder(v);
    }

    @Override
    public void onBindViewHolder(TopicViewHolder holder, int position) {
        final Tag tag = mFilteredTags.get(position);
        if (DEBUG) Timber.d("%s %d", tag, position);
        String text = String.format("%s (%d/%d)", tag.getText(), tag.getQuantity(), tag.getLearned());
        holder.topic.setText(text);
        holder.checkbox.setChecked(tag.isChecked());
        holder.checkbox.setOnClickListener(v -> {
            boolean isChecked = ((CheckBox)v).isChecked();
            //tag.setChecked(isChecked);
            // synchronize with all tags
            for (Tag t: mMasterCopyTags) {
                if (t.getId() == tag.getId()){
                    t.setChecked(isChecked);
                }
            }
            if (DEBUG) Timber.d("change %s", tag);
        });
    }

}

File: app/src/main/java/doit/study/droid/adapters/TopicsAdapter.kt
package doit.study.droid.adapters

import android.content.Context
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.TextView

import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView

import java.util.ArrayList

import doit.study.droid.R
import doit.study.droid.data.Tag
import timber.log.Timber

class TopicsAdapter : RecyclerView.Adapter<TopicsAdapter.TopicViewHolder>() {
    private var filteredTags: MutableList<Tag>? = null
    var tags: List<Tag>? = null
        set(tags) {
            field = tags
            filteredTags = ArrayList(tags)
            notifyDataSetChanged()
        }


    fun animateTo(models: List<Tag>) {
        if (DEBUG) Timber.d("Anim Filtered %d, current: %d", models.size, filteredTags!!.size)
        applyAndAnimateRemovals(models)
        if (DEBUG) Timber.d("Rem Filtered %d, current: %d", models.size, filteredTags!!.size)
        applyAndAnimateAdditions(models)
        if (DEBUG) Timber.d("Add Filtered %d, current: %d", models.size, filteredTags!!.size)
        applyAndAnimateMovedItems(models)
        if (DEBUG) Timber.d("Move Filtered %d, current: %d", models.size, filteredTags!!.size)
    }

    private fun applyAndAnimateRemovals(newTags: List<Tag>) {
        for (i in filteredTags!!.indices.reversed()) {
            val tag = filteredTags!![i]
            if (!newTags.contains(tag)) {
                removeItem(i)
            }
        }
    }

    private fun applyAndAnimateAdditions(newModels: List<Tag>) {
        var i = 0
        val count = newModels.size
        while (i < count) {
            val model = newModels[i]
            if (!filteredTags!!.contains(model)) {
                addItem(i, model)
            }
            i++
        }
    }

    private fun applyAndAnimateMovedItems(newModels: List<Tag>) {
        for (toPosition in newModels.indices.reversed()) {
            val model = newModels[toPosition]
            val fromPosition = filteredTags!!.indexOf(model)
            if (fromPosition >= 0 && fromPosition != toPosition) {
                moveItem(fromPosition, toPosition)
            }
        }
    }

    fun removeItem(position: Int): Tag {
        val model = filteredTags!!.removeAt(position)
        notifyItemRemoved(position)
        notifyItemRangeChanged(position, itemCount)
        return model
    }

    fun addItem(position: Int, model: Tag) {
        filteredTags!!.add(position, model)
        notifyItemInserted(position)
    }

    fun moveItem(fromPosition: Int, toPosition: Int) {
        val model = filteredTags!!.removeAt(fromPosition)
        filteredTags!!.add(toPosition, model)
        notifyItemMoved(fromPosition, toPosition)
    }

    class TopicViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal var topic: TextView = itemView.findViewById(R.id.topic_name)
        internal var checkbox: CheckBox = itemView.findViewById(R.id.checkbox_tag)
    }

    override fun getItemCount(): Int = filteredTags?.size ?: 0

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TopicViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.adapter_topic_item, parent, false)
        return TopicViewHolder(v)
    }

    override fun onBindViewHolder(holder: TopicViewHolder, position: Int) {
        val tag = filteredTags!![position]
        if (DEBUG) Timber.d("%s %d", tag, position)
        val text = String.format("%s (%d/%d)", tag.name, tag.questionsCounter, tag.questionsStudied)
        holder.topic.text = text
        holder.checkbox.isChecked = tag.selectionStatus
        holder.checkbox.setOnClickListener { v ->
            val isChecked = (v as CheckBox).isChecked
            tag.setChecked(isChecked)
            // synchronize with all tags
            for (t in tags!!) {
                if (t.id == tag.id) {
                    t.setChecked(isChecked)
                }
            }
            if (DEBUG) Timber.d("change %s", tag)
        }
    }


    companion object {
        private const val DEBUG = false
    }
}
